include/boost/geometry/algorithms/detail/touches/implementation.hpp

```diff
@@ -199,51 +199,30 @@ struct areal_interrupt_policy
     }
 };
 
-template<typename Geometry, typename PointInRingStrategy>
-struct check_each_ring_for_within
+template <typename Geometry1, typename Geometry2, typename Strategy>
+inline bool point_on_border_within(Geometry1 const& geometry1,
+                                   Geometry2 const& geometry2,
+                                   Strategy const& strategy)
 {
-    bool has_within;
-    Geometry const& m_geometry;
-    PointInRingStrategy const& m_strategy;
-
-    inline check_each_ring_for_within(Geometry const& g, PointInRingStrategy const& strategy)
-        : has_within(false)
-        , m_geometry(g)
-        , m_strategy(strategy)
-    {}
-
-    template <typename Range>
-    inline void apply(Range const& range)
-    {
-        typename geometry::point_type<Range>::type p;
-        geometry::point_on_border(p, range);
-        if ( !has_within && geometry::within(p, m_geometry, m_strategy) )
-        {
-            has_within = true;
-        }
-    }
-};
+    typename geometry::point_type<Geometry1>::type pt;
+    return geometry::point_on_border(pt, geometry1)
+        && geometry::within(pt, geometry2, strategy);
```

Comment:
A theoretical question, does anyone know if optimisers can handle this old-fashioned piece better?

```cpp
if  (geometry::point_on_border(pt, geometry1)) // pt is output
     return geometry::within(pt, geometry2, strategy); // pt is input
else
    return false;
```

I remember old compilers/optimisers tripping on some uses of ternary operator :)

Reply:
I don't know, so the version with ternary operator would be yet another one:
```
return geometry::point_on_border(pt, geometry1)
     ? geometry::within(pt, geometry2, strategy)
     : false;
```

I always assumed the compiler can do whatever it wants in any case and if we wanted to be sure we would have to check each version of each one of them (a good example is `switch` and which `case` is checked first, etc.). So I tend to use logical `&&` to check conditions one by one because it is the shortest.

Reply:
> I tend to use logical && to check conditions one by one because it is the shortest.

Yes, makes sense, shorter seems better again.
I'm too lazy to check the Compiler Explorer output right now.

>  version with ternary operator would be yet another one

Generally, I try to avoid the ternary operator. I'd rather use if-else instead.